ast/env.go

```diff
@@ -253,6 +270,12 @@ func (n *typeTreeNode) Put(path Ref, tpe types.Type) {
 	curr.value = tpe
 }
 
+func (n *typeTreeNode) PutArg(path Ref, i int, tpe types.Type) {
+	pos := json.Number(strconv.Itoa(i))
```

Comment:
I get the concern that we use refs to refer to functions, but their refs are slightly special in that they live outside the `data` namespace. This might help with the ambiguity issues, in that they aren't 'data' documents